src/main/java/world/bentobox/bentobox/api/flags/AbstractFlagListener.java

```diff
@@ -158,25 +196,41 @@ public boolean checkIsland(Event e, Location loc, Flag flag, boolean silent) {
 
         if (island.isPresent()) {
             // If it is not allowed on the island, "bypass island" moderators can do anything
-            if (!island.get().isAllowed(user, flag) && !user.hasPermission(getIWM().getPermissionPrefix(loc.getWorld()) + ".mod.bypass." + flag.getID() + ".island")) {
-                noGo(e, flag, silent);
-                // Clear the user for the next time
+            if (island.get().isAllowed(user, flag)) {
+                report(user, e, loc, flag,  Why.RANK_ALLOWED);
                 user = null;
-                return false;
-            } else {
+                return true;
+            } else if (user.hasPermission(getIWM().getPermissionPrefix(loc.getWorld()) + ".mod.bypass." + flag.getID() + ".island")) {
+                report(user, e, loc, flag,  Why.BYPASS_ISLAND);
                 user = null;
                 return true;
-            }
-        }
-        // The player is in the world, but not on an island, so general world settings apply
-        if (!flag.isSetForWorld(loc.getWorld())) {
+            } 
+            report(user, e, loc, flag,  Why.NOT_ALLOWED_ON_ISLAND);
             noGo(e, flag, silent);
+            // Clear the user for the next time
             user = null;
             return false;
-        } else {
+        }
+        // The player is in the world, but not on an island, so general world settings apply
+        if (flag.isSetForWorld(loc.getWorld())) {
+            report(user, e, loc, flag,  Why.ALLOWED_IN_WORLD);
             user = null;
             return true;
+        } else {
+            report(user, e, loc, flag,  Why.NOT_ALLOWED_IN_WORLD);
+            noGo(e, flag, silent);
+            user = null;
+            return false;
+        }
+    }
+
+    private void report(User user, Event e, Location loc, Flag flag, Why why) {
```

Comment:
Looks good. I was wondering if it was possible to send these reports directly ingame, but that may be a hard thing to do and that might also become a bit spammy.